cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java

```diff
@@ -803,6 +805,88 @@ public void suspendResumeSchedule(String namespaceId, String appId, String sched
   }
 
   /**
+   * Add a schedule for a program.
+   *
+   * @param programId the program id for which the schedule needs to be added
+   * @param scheduleSpec the schedule specification
+   * @throws NotFoundException when application is not found
+   * @throws SchedulerException on an exception when updating the schedule
+   */
+  public void addSchedule(ProgramId programId, ScheduleSpecification scheduleSpec)
+    throws NotFoundException, SchedulerException, AlreadyExistsException {
+    ApplicationSpecification appSpec = store.getApplication(programId.getParent());
+    if (appSpec == null) {
+      throw new ApplicationNotFoundException(programId.getParent());
+    }
+
+    String scheduleName = scheduleSpec.getSchedule().getName();
+    ScheduleSpecification existingScheduleSpec = appSpec.getSchedules().get(scheduleName);
+    if (existingScheduleSpec != null) {
+      throw new AlreadyExistsException(
+        new ScheduleId(programId.getNamespace(), programId.getApplication(), scheduleName));
+    }
+
+    // TODO: Figure out if there a way to make the scheduler update and store update transactional
+    scheduler.schedule(programId, scheduleSpec.getProgram().getProgramType(), scheduleSpec.getSchedule(),
+                       scheduleSpec.getProperties());
+    store.addSchedule(programId, scheduleSpec, false);
+  }
+
+  /**
+   * Update the schedule for a program.
+   *
+   * @param programId the program id for which the schedule needs to be updated
+   * @param scheduleSpecUpdate update to the schedule specification
+   * @throws NotFoundException when application is not found
+   * @throws SchedulerException on an exception when updating the schedule
+   */
+  public void updateSchedule(ProgramId programId, ScheduleSpecification scheduleSpecUpdate)
+    throws NotFoundException, SchedulerException {
+    ApplicationSpecification appSpec = store.getApplication(programId.getParent());
+    if (appSpec == null) {
+      throw new ApplicationNotFoundException(programId.getParent());
+    }
+
+    String scheduleName = scheduleSpecUpdate.getSchedule().getName();
+    ScheduleSpecification existingScheduleSpec = appSpec.getSchedules().get(scheduleName);
+    if (existingScheduleSpec == null) {
+      throw new NotFoundException(new ScheduleId(programId.getNamespace(), programId.getApplication(), scheduleName));
+    }
+
+    // TODO: Overlay the schedule spec update on to the existing spec
+    // TODO: Also handle schedule type change
```

Comment:
In my opinion that is not needed. One can always delete the existing schedule and then add a new one